### PR TITLE
Dynamic Mapping Server Creation

### PR DESCRIPTION
**Pull request by scalbertm**
First commit, adding a minimal version of the server.
Includes:
- Mapping Storage and Creation
- Conversion of a Mapping to a groovy Script
- Scripts basic API

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    <artifactId>gridsuite-mapping-server</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+
+    <packaging>jar</packaging>
+    <name>Mapping server</name>
+    <description>Mapping service</description>
+    <url>http://www.gridsuite.org/</url>
+<!--
+    <scm>
+        <connection>scm:git:https://github.com/gridsuite/actions-server.git</connection>
```
Comment (jonenst) [inline pom.xml:32]: mapping-server

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            <artifactId>spring-data-jpa</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-webflux</artifactId>
+            <version>5.2.8.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
```
Comment (jonenst) [inline pom.xml:199]: this one is probably a mistake ?

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            <version>RELEASE</version>
+            <scope>compile</scope>
+        </dependency>
+    </dependencies>
+
+    <profiles>
+        <!-- needed for embedded-cassandra to download cassandra (only once) through authentified proxies -->
```
Comment (jonenst) [inline pom.xml:206]: cassandra not used in this project

File: src/main/java/org/gridsuite/mapping/server/MappingApplication.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static void main(String[] args) {
+        SpringApplication.run(MappingApplication.class, args);
+    }
+
+    @Bean
+    public WebMvcConfigurer corsConfigurer() {
+        return new WebMvcConfigurer() {
+            @Override
+            public void addCorsMappings(CorsRegistry registry) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/MappingApplication.java:32]: why cors ? we don't use it in our other microservices. We reverse proxy the domain hosting the javascript instead

**Comment by jonenst** [comment]
missing license header from many java files

File: src/main/java/org/gridsuite/mapping/server/MappingApplication.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+import org.springframework.web.servlet.config.annotation.CorsRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/MappingApplication.java:17]: please update the author information

File: src/main/java/org/gridsuite/mapping/server/controller/MappingController.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    @ApiResponses(value = {@ApiResponse(code = 200, message = "The list of mappings")})
+    public ResponseEntity<List<InputMapping>> getMappingList() {
+        return ResponseEntity.ok().contentType(MediaType.APPLICATION_JSON).body(mappingService.getMappingList());
+    }
+
+    @PostMapping(value = "/{mappingName}")
+    @ApiOperation(value = "Create a mapping")
+    @ApiResponses(value = {@ApiResponse(code = 200, message = "The id of the mapping"),
+            @ApiResponse(code = 409, message = "The mapping already exist"),
+            @ApiResponse(code = 500, message = "The storage is down or a file with the same name already exists")})
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/controller/MappingController.java:33]: is this a copy pasted error message ? if so, update it.

File: src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Repository
+public interface InstanceModelRepository extends JpaRepository<InstanceModelEntity, UUID> {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java:13]: Why UUID in the template arguments ? InstanceModelEntity has @Id on a String

File: src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+    List<InstanceModelEntity> findAll();
+
+    Optional<InstanceModelEntity> findById(String id);
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java:17]: Not needed if you have the correct template parameters?

File: src/main/java/org/gridsuite/mapping/server/repository/MappingRepository.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.gridsuite.mapping.server.model.MappingEntity;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Repository
+public interface MappingRepository extends JpaRepository<MappingEntity, UUID> {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/repository/MappingRepository.java:14]: why UUID in the template parameters, should be String ?

File: src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    @Getter
+    @Setter
+    public class SortedMapping implements Mapping {
+        private String name;
+        private ArrayList<SortedRules> sortedRules;
+
+        public SortedMapping(InputMapping mapping) {
+            name = mapping.getName();
+            sortedRules = new ArrayList<SortedRules>();
+            HashMap<EquipmentType, ArrayList<Rule>> sortingRules = new HashMap<EquipmentType, ArrayList<Rule>>();
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java:96]: use java7's type inference: new HashMap<>, newArrayList<>

File: src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            mapping.getRules().stream().forEach(rule -> {
+                EquipmentType ruleType = rule.getEquipmentType();
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java:98]: this whole blocks looks like collectors.groupingBy ?

File: src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    @AllArgsConstructor
+    public class FlatRule {
+        private EquipmentType equipmentType;
+
+        private InstanceModelEntity mappedModel;
+
+        private String composition;
+
+        public FlatRule(Rule rule) {
+            Optional<InstanceModelEntity> foundModel = instanceModelRepository.findById(rule.getMappedModel());
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/service/implementation/ScriptServiceImpl.java:150]: how many rules are expected here ? calling findById for each rule is going to be very long if there are many rules

File: src/main/resources/application.yaml [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      # Uncomment action to automatically generate the database sql schema file
+      javax:
+        persistence:
+          schema-generation:
+            scripts:
+              #action: create
+              create-target: src/main/resources/actions.sql
+              create-source: metadata
+
+network-store-server:
```
Comment (jonenst) [inline src/main/resources/application.yaml:37]: network-store is not used ?

File: src/main/resources/application.yaml [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  base-uri: http://localhost:8080
+
+logging:
+  level:
+    ROOT: INFO
+
+dbName: postgres
```
Comment (jonenst) [inline src/main/resources/application.yaml:44]: the postgres db is reserved, you must set your name here. Like dynamicmapping

File: src/main/java/org/gridsuite/mapping/server/model/FilterEntity.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Column(name = "type")
+    @Enumerated
+    private PropertyType type;
+
+    @Column(name = "operand")
+    @Enumerated
+    private Operands operand;
+
+    // Value is the value casted to a string for persistence issue
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/model/FilterEntity.java:39]: maybe implement the same system as https://github.com/gridsuite/network-modification-server/blob/master/src/main/java/org/gridsuite/modification/server/entities/elementary/BooleanElementaryModificationEntity.java ?

File: src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Repository
+public interface InstanceModelRepository extends JpaRepository<InstanceModelEntity, UUID> {
+
+    List<InstanceModelEntity> findAll();
```
[the remark this answers is not in the document]
Comment (scalbertm) [reply src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java:15]: Not needed for this POC but will be once instanciated models will be created user-side by merging parameters and dynamic models.

File: src/main/java/org/gridsuite/mapping/server/model/RuleEntity.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    @Enumerated
+    private EquipmentType equipmentType;
+
+    @Column(name = "model", nullable = false)
+    private String mappedModel;
+
+    @Column(name = "composition", nullable = false)
+    private String composition;
+
+    @OneToMany(targetEntity = FilterEntity.class, mappedBy = "ruleId", cascade = CascadeType.ALL, orphanRemoval = true)
```
[the remark this answers is not in the document]
Comment (scalbertm) [reply src/main/java/org/gridsuite/mapping/server/model/RuleEntity.java:36]: It works for the small sample I have tried.
There should not be calls from filter to rule so no need for bidirectional.
Used to be CollectionTable but could not make it work

**Review by geofjamg — CHANGES_REQUESTED** [review]
I didn't look at the code yet but some general comments:
 - every files need to copyright RTE and authorship in the class javadoc
 - Code coverage is very low, we usually require 80%
 -

File: pom.xml [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/commons-io/commons-io -->
```
Comment (jonenst) [inline pom.xml:111]: remove this comment

File: pom.xml [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-webflux</artifactId>
+            <version>5.2.8.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot</artifactId>
+            <version>2.2.9.RELEASE</version>
```
Comment (jonenst) [inline pom.xml:176]: don't put the version here

File: src/main/java/org/gridsuite/mapping/server/dto/filters/AbstractFilter.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+        @JsonSubTypes.Type(value = StringFilter.class, name = "STRING"),
+        @JsonSubTypes.Type(value = BooleanFilter.class, name = "BOOLEAN")})
+@Data
+public abstract class AbstractFilter {
+    private String filterId;
+    private String property;
+    private Operands operand;
+
+    public static AbstractFilter createFilterFromEntity(FilterEntity filterEntity) {
+        AbstractFilter filter = new StringFilter();
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/dto/filters/AbstractFilter.java:34]: don't create a new StringFilter here, instead put a default in the switch statement and throw an exception for unknown types

File: src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jonenst) [reply src/main/java/org/gridsuite/mapping/server/repository/InstanceModelRepository.java:23]: no I mean JpaRepository already has this method so redeclaring here has no effect

File: src/main/java/org/gridsuite/mapping/server/utils/Methods.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public static ArrayList<String> convertStringToList(String stringArray) {
+        ArrayList<String> converted = new ArrayList();
+        converted.addAll(Arrays.asList(stringArray.split("\\s*,\\s*")));
+        return converted;
+    }
+
+    public static String convertListToString(List<String> array) {
+        return array.stream().reduce("", (acc, element) -> {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/mapping/server/utils/Methods.java:29]: use string joiner ? or at least a stringbuilder to avoid o(n²) performance

**Review by jonenst — APPROVED** [review]

**Pull request merged**